src/Microsoft.Extensions.Configuration.AzureAppConfiguration/DomainVerifier.cs

```diff
@@ -0,0 +1,145 @@
+            if (string.IsNullOrEmpty(srvHostName))
+            {
+                return string.Empty;
+            }
+
+            IEnumerable<string> validDomains = await GetValidDomainsFromTlsCert(srvHostName).ConfigureAwait(false);
+            
+            foreach (var domain in validDomains)
+            {
+                if (originEndpoint.DnsSafeHost.EndsWith(domain, StringComparison.OrdinalIgnoreCase))
+                {
+                    return domain;
+                }
+            }
+
+            return string.Empty;
+        }
+
+        private static async Task<List<string>> GetValidDomainsFromTlsCert(string endpoint)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(endpoint));
+
+            var validDomains = new List<string>();
+
+            X509Certificate2 cert = null;
+
+            using (var client = new TcpClient(endpoint, TlsPort))
+            using (var sslStream = new SslStream(client.GetStream(), leaveInnerStreamOpen: false))
+            {
+
+                // Initiate the connection, so it will download the server certificate
+                await sslStream.AuthenticateAsClientAsync(endpoint).ConfigureAwait(false);
+
+                var serverCertificate = sslStream.RemoteCertificate;
+
+                if (serverCertificate == null)
+                {
+                    return validDomains;
+                }
+
+                cert = new X509Certificate2(serverCertificate);
+            }
+
+            if (cert == null)
+            {
+                return validDomains;
+            }
+
+            // Get the Subject Alternative Name (SAN) extension
+            var sanExtension = cert.Extensions[SubjectAltNameOid];
+
+            if (sanExtension != null)
+            {
+                IEnumerable<string> formattedSanExtensions = sanExtension
+                    .Format(true)
+                    .Split(new[] { Environment.NewLine }, StringSplitOptions.RemoveEmptyEntries);
+
+                foreach (string formattedExtension in formattedSanExtensions)
+                {
+                    // Valid pattern should be "DNS Name=*.domain.com"
+                    string[] parts = formattedExtension.Split('=');
+
+                    if (parts.Length != 2)
+                    {
+                        continue;
+                    }
+
+                    string value = parts[1];
+
+                    // Skip non-multi domain
+                    if (!value.StartsWith(MultiDomainWildcard))
+                    {
+                        continue;
+                    }
+
+                    // .domain.com
+                    string domain = value.Substring(1);
+
+                    if (domain.Length > 1 && !validDomains.Contains(domain))
+                    {
+                        validDomains.Add(domain);
+                    }
+                }
+            }
+
+            // Get the Common Name (CN) from the Subject Name if Subject Alternative Name is not available

```

Comment:
All of our certs use SAN right?

Reply:
They should.

Reply:
then we can remove this.

Reply:
This is a way recommend by a RFC doc, but given we ensure we use SAN, we can remove this.